golem-worker-service-base/src/parser/place_holder_parser.rs

```diff
@@ -11,13 +11,69 @@ pub struct PlaceHolder(pub String);
 impl GolemParser<PlaceHolder> for PlaceHolderPatternParser {
     fn parse(&self, str: &str) -> Result<PlaceHolder, ParseError> {
         match parse_place_holder(str) {
-            Ok(value) => Ok(value.1),
+            Ok((_, parsed)) => Ok(PlaceHolder(parsed.to_string())),
             Err(err) => Result::Err(ParseError::Message(err.to_string())),
         }
     }
 }
 
-pub fn parse_place_holder(input: &str) -> IResult<&str, PlaceHolder> {
-    delimited(tag("{"), take_while(|c| c != '}'), tag("}"))(input)
-        .map(|(rest, captured)| (rest, PlaceHolder(captured.to_string())))
+pub fn parse_place_holder(input: &str) -> IResult<&str, &str> {
+    delimited(tag("{"), take_until_unbalanced('{', '}'), tag("}"))(input)
+}
+
+// https://stackoverflow.com/questions/70630556/parse-allowing-nested-parentheses-in-nom
+fn take_until_unbalanced(
+    opening_bracket: char,
+    closing_bracket: char,
+) -> impl Fn(&str) -> IResult<&str, &str> {
+    move |i: &str| {
+        let mut index = 0;
+        let mut bracket_counter = 0;
+        while let Some(n) = &i[index..].find(&[opening_bracket, closing_bracket, '\\'][..]) {
+            index += n;
+            let mut it = i[index..].chars();
+            match it.next().unwrap_or_default() {
+                '\\' => {
+                    // Skip the escape char `\`.
+                    index += '\\'.len_utf8();
+                    // Skip also the following char.
+                    let c = it.next().unwrap_or_default();
+                    index += c.len_utf8();
+                }
+                c if c == opening_bracket => {
+                    bracket_counter += 1;
+                    index += opening_bracket.len_utf8();
+                }
+                c if c == closing_bracket => {
+                    // Closing bracket.
+                    bracket_counter -= 1;
+                    index += closing_bracket.len_utf8();
+                }
+                // Can not happen.
+                _ => unreachable!(),
+            };
+            // We found the unmatched closing bracket.
+            if bracket_counter == -1 {
+                // We do not consume it.
+                index -= closing_bracket.len_utf8();
+                return Ok((&i[index..], &i[0..index]));
```

Comment:
Lots of slice operations. I think we can avoid it, if this is about finding  `query{}` out of `{query{}}`

Reply:
I think fixing the first delimiter, and see if the middle portion is unbalanced, before the last delimiter looks like we are squeezing the usecase back to `nom`. 


 I remember using `nom` when I started learning `Rust` last year, and honestly the API is a pain for me. :D

Reply:
May not handle escape, but  i was writing it using my simple tokenizer. Not saying to use it (because I am removing cursor soon), but just FYI


```scala

    #[test]
    fn test_path_pattern() {
        let mut cursor = Tokenizer::new("{variable{}}").run().to_cursor();

        match  cursor.next_token() {
            Some(Token::LCurly) => {
                let result = cursor.capture_string_until(vec![&Token::LCurly], &Token::RCurly).unwrap();
                assert_eq!(result, "variable{}");
            }
            _ => panic!("Expected LCurly"),
        }
    }
```